File: Mage/MageAuthAPI.m
//
//  MageAuthAPI.m
//  MAGE
//
//  Created by Brent Michalski on 9/5/25.
//  Copyright © 2025 National Geospatial Intelligence Agency. All rights reserved.
//

#import "MageAuthAPI.h"
#import "MageSessionManager.h"
//#import "MAGE-Bridging-Header.h"
#import "MAGE-Swift.h"

@implementation MageAuthAPI

+ (void)changePasswordWithCurrent:(NSString *)current
                      newPassword:(NSString *)newPassword
               confirmNewPassword:(NSString *)confirmNewPassword
                       completion:(void (^)(NSHTTPURLResponse * _Nullable, NSData * _Nullable, NSError * _Nullable))completion
{
    NSURL *base = [MageServer baseURL];
    if(!base) {
        if (completion) completion(nil, nil, [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorBadURL userInfo:@{NSLocalizedDescriptionKey:@"Missing base URL"}]);
        return;
    }
    
    NSString *urlString = [NSString stringWithFormat:@"%@/%@", base.absoluteString, @"api/users/myself/password"];
    
    NSDictionary *parameters = @{
        @"password": current ?: @"",
        @"newPassword": newPassword ?: @"",
        @"newPasswordConfirm": confirmNewPassword ?: @""
    };
    
    MageSessionManager *manager = [MageSessionManager sharedManager];
    
    NSMutableURLRequest *request = [manager.requestSerializer requestWithMethod:@"PUT"
                                                                      URLString:urlString
                                                                     parameters:parameters
                                                                          error:nil];
    
    NSURLSessionDataTask *task = [manager dataTaskWithRequest:request
                                               uploadProgress:nil
                                             downloadProgress:nil
                                            completionHandler:^(NSURLResponse * _Nonnull response, id _Nullable responseObject, NSError * _Nullable error)
                                  {
        if (completion) {
            NSHTTPURLResponse *http = (NSHTTPURLResponse *)response;
            NSData *body = error.userInfo[AFNetworkingOperationFailingURLResponseDataErrorKey];
            completion(http, body, error);
        }
    }];
    
    [manager addTask: task];
}


+ (void)requestSignupCaptchaForUsername:(NSString *)username
                             background:(NSString *)backgroundHex
                             completion:(void (^)(NSString * _Nullable,
                                                  NSString * _Nullable,
                                                  NSError * _Nullable))completion
{
    NSURL *base = [MageServer baseURL];
    if (!base) {
        if (completion) completion(nil, nil,
                                   [NSError errorWithDomain: NSURLErrorDomain code:NSURLErrorBadURL
                                                   userInfo:@{NSLocalizedDescriptionKey:@"Missing base URL"}]);
        return;
    }
    
    NSString *urlString = [NSString stringWithFormat:@"%@/%@", base.absoluteString, @"api/users/signups"];
    NSDictionary *params = @{ @"username": username ?: @"", @"background": backgroundHex ?: @"#FFFFFF" };
    
    MageSessionManager *manager = [MageSessionManager sharedManager];
    NSMutableURLRequest *req = [manager.requestSerializer requestWithMethod:@"POST"
                                                                  URLString:urlString
                                                                 parameters:params
                                                                      error:nil];
    
    NSURLSessionDataTask *task = [manager dataTaskWithRequest:req uploadProgress:nil downloadProgress:nil
                                            completionHandler:^(NSURLResponse * _Nonnull response,
                                                                id  _Nullable responseObject,
                                                                NSError * _Nullable error)
                                  {
        void (^finish)(NSString*, NSString*, NSError*) = ^(NSString *t, NSString *c, NSError *e) {
            // Hop to main so SwiftUI state updates are safe
            dispatch_async(dispatch_get_main_queue(), ^{ if (completion) completion(t, c, e); });
        };
        
        if (error) { finish(nil, nil, error); return ; }
        
        id obj = responseObject;
        
        if([obj isKindOfClass:[NSData class]]) {
            NSError *jsonErr = nil;
            id parsed = [NSJSONSerialization JSONObjectWithData:(NSData *)obj options:0 error:&jsonErr];
            if (jsonErr) { finish(nil, nil, jsonErr); return ; }
            obj = parsed ?: @{};
        }
        
        if (![obj isKindOfClass:[NSDictionary class]]) {
            NSError *e = [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorCannotParseResponse
                                         userInfo:@{NSLocalizedDescriptionKey:@"Unexpected signup captcha response"}];
            finish(nil, nil, e);
            return;
        }

        NSDictionary *dict = (NSDictionary *)obj;
        NSString *token = [dict objectForKey:@"token"];
        NSString *captcha = [dict objectForKey:@"captcha"];
        
        if (token.length == 0 || captcha.length == 0) {
            NSError *e = [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorCannotParseResponse
                                         userInfo:@{NSLocalizedDescriptionKey:@"Missing captcha or token"}];
            finish(nil, nil, e);
            return;
        }
        finish(token, captcha, nil);
    }];
    
    [manager addTask:task];
}

+ (void)completeSignupWithParameters:(NSDictionary *)parameters
                               token:(NSString *)token
                          completion:(void (^)(NSHTTPURLResponse * _Nullable http,
                                               NSData * _Nullable errorBody,
                                               NSError * _Nullable error))completion
{
    NSURL *base = [MageServer baseURL];
    if (!base) { if (completion) completion(nil, nil,
                                            [NSError errorWithDomain: NSURLErrorDomain code:NSURLErrorBadURL
                                                            userInfo:@{NSLocalizedDescriptionKey:@"Missing base URL"}]); return; }
    
    NSString *urlString = [NSString stringWithFormat:@"%@/%@", base.absoluteString, @"api/users/signups/verifications"];
    
    MageSessionManager *manager = [MageSessionManager sharedManager];
    NSMutableURLRequest *req = [manager.requestSerializer requestWithMethod:@"POST"
                                                                  URLString:urlString
                                                                 parameters:parameters
                                                                      error:nil];
    [req setValue:[NSString stringWithFormat:@"Bearer %@", token ?: @""] forHTTPHeaderField:@"Authorization"];
    
    NSURLSessionDataTask *task = [manager dataTaskWithRequest:req uploadProgress:nil downloadProgress:nil
                                            completionHandler:^(NSURLResponse * _Nonnull response,
                                                                id  _Nullable responseObject,
                                                                NSError * _Nullable error)
                                  {
        NSHTTPURLResponse *http = (NSHTTPURLResponse *)response;
        NSData *body = error.userInfo[AFNetworkingOperationFailingURLResponseDataErrorKey];
        if (completion) completion(http, body, error);
    }];

    [manager addTask: task];
}

@end

File: Mage/Services/MageAuthAPI.swift
//
//  MageAuthAPI.swift
//  MAGE
//
//  Created by Brent Michalski on 9/17/25.
//  Copyright © 2025 National Geospatial Intelligence Agency. All rights reserved.
//

import Foundation

@objc public final class MageAuthAPI: NSObject {
    
    // MARK: - CAPTCHA
    
    // GET /api/users/signups/verifications?username=&background=
    // Completion: (token, imageBase64, error)
    @objc public class func getSignupCaptcha(
        forUsername username: String,
        background: String,
        completion: @escaping (_ token: String?, _ base64: String?, _ error: NSError?) -> Void
    ) {
        
        let url = "/api/users/signups/verifications"
        let params: [String: Any] = [
            "username": username,
            "background": background
        ]
        
        let sessionManager = MageSessionManager.shared()

        let task = sessionManager?.get_TASK(
            url,
            parameters: params,
            progress: nil,
            success: { (task: URLSessionDataTask, response: Any?) in
                if let data = response as? Data {
                    if let json = (try? JSONSerialization.jsonObject(with: data, options: [])) as? [String: Any] {
                        let (tok, b64) = Self.parseCaptchaJSON(json)
                        completion(tok, b64, nil)
                    } else {
                        completion(nil, nil, Self.makeError(0, "Unexpected captcha response"))
                    }
                } else if let json = response as? [String: Any] {
                    let (tok, b64) = Self.parseCaptchaJSON(json)
                    completion(tok, b64, nil)
                } else {
                    completion(nil, nil, Self.makeError(0, "Unexpected captcha response"))
                }
            },
            failure: { (task: URLSessionDataTask?, error: Error) in
                completion(nil, nil, error as NSError)
            }
        )
        
        if let task { sessionManager?.addTask(task) }
    }
    
    private static func parseCaptchaJSON(_ json: [String: Any]) -> (String?, String?) {
        let token = (json["token"] as? String)
        ?? (json["verificationToken"] as? String)
        ?? (json["capthchaToken"] as? String)
        
        let base64 = (json["captcha"] as? String)
        ?? (json["image"] as? String)
        ?? (json["imageBase64"] as? String)
        
        return (token, base64)
    }
    
    
    // MARK: - SIGNUP
    
    /// POST /api/users/signups
    /// Expects the signup body + captcha token/text
    @objc public class func signup(
        withParameters parameters: [String: Any],
        captchaText: String,
        token: String,
        completion: @escaping (_ http: HTTPURLResponse?, _ error: NSError?) -> Void) {
            let url = "/api/users/signups"
            
            var body = parameters
            
            // Try multiple key names to match common backends (adjust to match yours exactly)
            body["captcha"] = captchaText
            body["captchaText"] = captchaText
            body["token"] = token
            body["verificationToken"] = token
            
            let sessionManager = MageSessionManager.shared()
            
            let task = sessionManager?.post_TASK(url, parameters: body, progress: nil, success: { task, _ in
                completion(task.response as? HTTPURLResponse, nil)
            }, failure: { task, error in
                completion(task?.response as? HTTPURLResponse, error as NSError)
            })
            
            if let task { sessionManager?.addTask(task) }
        }

    
    // PUT/POST change password (path can be adjusted to match server)
    @objc public class func changePassword(
        currentPassword: String,
        newPassword: String,
        confirmedPassword: String,
        completion: @escaping(_ http: HTTPURLResponse?, _ error: NSError?) -> Void
    ) {
        guard let base = MageServer.baseURL()?.absoluteString else {
            completion(nil, NSError(domain: NSURLErrorDomain, code: NSURLErrorBadURL, userInfo: [NSLocalizedDescriptionKey: "Missing base URL"]))
            return
        }
        
        let url = "\(base)/api/users/me/password"
        let body: [String: Any] = [
            "currentPassword": currentPassword,
            "password": newPassword,
            "passwordconfirm": confirmedPassword
        ]
        
        let sessionManager = MageSessionManager.shared()
        // If you don't have put_TASK, using post_TASK works too if server accepts POST.
        let task = sessionManager?.put_TASK(url, parameters: body, success: { task, _ in
            completion(task.response as? HTTPURLResponse, nil)
        }, failure: { task, error in
            completion(task?.response as? HTTPURLResponse, error as NSError)
        }) ?? sessionManager?.post_TASK(url, parameters: body, progress: nil, success: { task, _ in
            completion(task.response as? HTTPURLResponse, nil)
        }, failure: { task, error in
            completion(task?.response as? HTTPURLResponse, error as NSError)
        })
        
        if let task { sessionManager?.addTask(task) }
    }

    
    // MARK: - Utils & Helpers
    private static func makeError(_ code: Int, _ message: String) -> NSError {
        NSError(domain: "MAGE.Auth", code: code, userInfo: [NSLocalizedDescriptionKey: message])
    }
}


/// OLD: - delete before release
//@objc public class func completeSignup(
//    withParameters params: [String: Any],
//    token: String,
//    completion: @escaping (_ http: HTTPURLResponse?, _ body: Data?, _ error: NSError?) -> Void
//) {
//    guard let base = MageServer.baseURL()?.absoluteString else {
//        completion(nil, nil, NSError(domain: NSURLErrorDomain, code: NSURLErrorBadURL,
//                                     userInfo: [NSLocalizedDescriptionKey: "Missing base URL"]))
//        return
//    }
//    
//    let url = "\(base)/api/users/signups/verifications"
//    
//    // If your server expects the token in headers, add it here. If it expects it in the body, you're already passing it.
//    let sessionManager = MageSessionManager.shared()
//    let task = sessionManager?.post_TASK(url, parameters: params, progress: nil, success: { task, response in
//        completion(task.response as? HTTPURLResponse, response as? Data, nil)
//    }, failure: { task, error in
//        completion(task?.response as? HTTPURLResponse, nil, error as NSError)
//    })
//    if let task { sessionManager?.addTask(task) }
//}
